blog/2024-11-18-relay-ops.md

```diff
@@ -0,0 +1,60 @@
+
+## Relay Upgrade
+
+The current Bluesky relay instance is available at `wss://bsky.network`. A more specific hostname for this relay instance is (XXX: public old relay hostname).
+
+Instead of upgrading the relay in-place, we are going to start a new relay from scratch, which means event stream ordering will be different and sequence numbers will not align with the current relay. This new instance will run at (XXX: public new relay hostname).
+
+There will be at least a short window when both relays are running, so downstream consumers can update their services to "cut over". At some point soon after the new relay is deployed, we will switch the generic `bsky.network` hostname to point at it.
+
+What firehose consumers need to do about the cut over depends on how careful they need to be about processing every single event on the firehose.
+
+Services which always connect to the firehose from the current offset don’t need to do anything. This includes developer tools like `goat`, and real-time sampling like Firesky.
+
+Best-effort consumers can likely be configure to consume from the current firehose offset, and simply restart, missing at most a few seconds of events. This probably makes sense for feed generators, analytics, automated labeling, etc.
+
+Services which want to really minimize the number of missed events can take a different approach. They can be programmed to connect to the new relay with no cursor, detect the current sequence number, then reconnect with a cursor a couple minutes back (at current rates this means a few hundred thousand events back). This results in a time overlap window between the two relays (with some events processed twice), which helps minimize the chance of missing any events.
+
+## Firehose Fan-out and Rainbow
+
+Firehose fan-out servers have a single upstream connection to a relay firehose, and re-broadcast that stream to multiple subscribing clients. They maintain a local backfill window (allowing re-connection with a cursor), but do not implement endpoints like `getRecord` or `getRepo`, only `com.atproto.sync.subscribeRepos`. They do not re-validate the event stream (eg, don't verify signatures). They maintain the exact sequence numbers of the upstream relay, meaning that they are interchangeable with the relay and with sibling fan-out servers, and can be placed behind a load-balancer.
+
+Today we are releasing Rainbow, a fan-out service implemented in Go. The [source code is available](https://github.com/bluesky-social/indigo/tree/main/cmd/rainbow) in the `indigo` git repository, with the same MIT/Apache licensing as our other open source projects.
+
+In the future, we may run additional Rainbow instances outside our primary data centers, similar to how we offer Jetstream instances today. This distributes the networking over more routers, and could improve connection reliability and throughput for subscribers in regions outside North America.
```

Comment:
```suggestion
In the future, we may run additional Rainbow instances outside our primary data centers, similar to how we offer Jetstream instances today. This distributes the network over more routers, and could improve connection reliability and throughput for subscribers in regions outside North America.
```
`networking` or `network`, not sure which is correct here?

Reply:
I think "networking" is clearer? but neither are great